packages/theme/src/cli/utilities/theme-environment/hot-reload/client.ts

```diff
@@ -74,38 +112,118 @@ function hotReloadScript() {
     }
   }
 
-  const refreshSections = async (data: HotReloadEvent & {type: 'section' | 'extAppBlock'}, elements: Element[]) => {
+  const buildSectionHotReloadUrl = (sectionId: string, data: HotReloadFileEvent) => {
+    if (isLocalPreview) {
+      // Note: Change this to mimic SFR API in CLI
+      const prefix = data.payload?.isAppExtension ? 'app' : 'section'
+      const params = [
+        `section-id=${encodeURIComponent(sectionId)}`,
+        `${prefix}-template-name=${encodeURIComponent(data.key)}`,
+        `pathname=${encodeURIComponent(window.location.pathname)}`,
+        `search=${encodeURIComponent(window.location.search)}`,
+      ].join('&')
+
+      return `${hotReloadOrigin}/__hot-reload/render?${params}`
+    }
+
+    const url = window.location.pathname
+    const params = new URLSearchParams({
+      _fd: '0',
+      pb: '0',
+    })
+
+    for (const [key, value] of new URLSearchParams(window.location.search)) {
+      params.append(key, value)
+    }
+
+    // The Section Rendering API takes precendence over the Block Rendering API.
+    if (sectionId) {
+      params.append('section_id', sectionId)
+    }
+
+    return `${url}?${params}`
+  }
+
+  const oseActions = {
+    startDataReload: async (signal: AbortSignal) => {
+      if (!isOSE) return null
+
+      // Force OSE to show the loading state
+      window.dispatchEvent(new Event('pagehide'))
+
+      return fetch(window.location.href, {
+        // Note: enable these properties when we have access to replace_templates
+        // method: 'POST',
+        // body: storefrontReplaceTemplatesParams(data.replaceTemplates),
+        // This is required to get the OnlineStoreEditorData script:
+        headers: {Accept: 'text/html'},
+        signal,
+      })
+        .then((response) => response.text())
+        .catch((error) => {
+          logError('Error fetching full page reload for section settings', error)
+          return null
+        })
+    },
+    finishDataReload: async (oseDataPromise: Promise<string | null>) => {
+      if (!isOSE) return null
+
+      const refreshedHtml = await oseDataPromise
+      const newOSEData = new DOMParser()
+        .parseFromString(refreshedHtml ?? '', 'text/html')
+        .querySelector('#OnlineStoreEditorData')?.textContent
+
+      if (newOSEData) {
+        const oseDataElement = document.querySelector('#OnlineStoreEditorData')
+        if (oseDataElement && newOSEData !== oseDataElement.textContent) {
+          oseDataElement.textContent = newOSEData
+          logInfo('OSE data updated')
+        }
+      }
+
+      // OSE reads the new data after the page is loaded
+      window.dispatchEvent(new Event('load'))
+    },
```

Comment:
We could remove these actions since they are not used yet. Using them is faster but it seems to create some small issues with OSE state.